[#65606] Apply description to links with blank target

myabc opened the pull request:
# Ticket
https://community.openproject.org/wp/65606 

# What are you trying to accomplish?

Uses a `ExternalLinksController` Stimulus controller to set `aria-describedby` on all links with `target="blank"`. Uses a `MutationObserver` to ensure links that are added to the DOM dynamically (e.g. via Turbo Drive) are also updated.

# What approach did you choose and why?

Explained in my comment below: https://github.com/opf/openproject/pull/20282#issuecomment-3304422827

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc committed:
[#65606] Apply description to links with blank target

Uses a `ExternalLinksController` Stimulus controller to set
`aria-describedby` on all links with `target="blank"`. Uses a
`MutationObserver` to ensure links that are added to the DOM dynamically
(e.g. via Turbo Drive) are also updated.

Co-Authored-By: Behrokh Satarnejad <62008897+bsatarnejad@users.noreply.github.com>

akabiru reviewed (commented):
Looks solid to me! 🌟

HDinger reviewed (commented):
The changes look good. 👏  I'd like to see a comment in controller explaining why this is needed. Otherwise, I see the danger of someone deleting that somewhen because they think it is unneeded. Also a test could help to prevent that.

bsatarnejad reviewed (approved):
LGTM 👍🏼

myabc committed:
Add a root eslint config

Attempt to fix spurious Copilot comments.

myabc committed:
Implement MutationHelper, extracting boilerplate

Moves most of the logic out of the Stimulus Controller and into the new
helper. Adds a small debounce to coalesce consecutive muation updates.

`MutationHelper` is conceived as aa declarative wrapper around
`MutationObserver` - however it remains to be seen how useful it will be
in other DOM change use cases.

myabc committed:
Document ExternalLinksController

myabc committed:
Add simple feature spec for external links

myabc committed:
Tell eslint to ignore base no-unused-vars rule

`@typescript-eslint/no-unused-vars` already extends this eslint rule.
However, the base rule must be disabled in order to avoid incorrect
errors (e.g. with TypeScript function types).

See https://typescript-eslint.io/rules/no-unused-vars/

myabc committed:
Append id to aria-describedby if existing value

myabc committed:
Use MutationHelper in ng service spec

myabc commented:
Hi @HDinger, @bsatarnejad and @akabiru,

I've made the following changes since you last looked at the PR:

- [X] added documentation (JSDoc comment) to the Stimulus controller to make it clear what it does.
- [X] added a very simple feature spec that checks all external links.
- [X] created a `MutationHelper` class and moved a lot of the mutation filtering boilerplate there.
- [X] not clobber existing `aria-describedby` values.

I used AI to help me with the `MutationHelper` class, and although I also modified (and cut down LOC), I'm still not happy with the results. It's over-engineered for what we need currently and the API not that much of an improvement over plain `MutationObserver`.

But, unfortunately I've run out of time before vacation.... if someone else is keen to refactor and improve, then please feel free!

myabc commented on frontend/src/stimulus/helpers/mutation-helper.ts:
```diff
@@ -0,0 +1,355 @@
+ * - Debounce/coalesce bursts of mutations.
+ * - Utility static helpers for common patterns (observeAdded, observeRemoved, observeOnce).
+ */
+export class MutationHelper<E extends Element> {
```
Perhaps this file should live under `frontend/src/app/shared/helpers` instead?

myabc commented on frontend/src/stimulus/helpers/mutation-helper.ts:
```diff
@@ -0,0 +1,355 @@
+    const ctx:ParentNode|null = (isElement(node) ? (node) : root) ?? null;
+
+    if (ctx) {
+      // If we have a selector, querySelectorAll is fast; otherwise walk children as Elements.
```
Cool, but ChatGPT v5 takes all the credit. This might be good if we're creating a library with benchmarks, but this is prematurely optimisation for our needs.

myabc commented on eslint.config.mjs:
```diff
@@ -0,0 +1,5 @@
+import frontendConfig from './frontend/eslint.config.mjs';
+
+export default [
+  ...frontendConfig
```
An attempt to quieten Copilot warnings about (valid) code formatting choices.

bsatarnejad commented:
@HDinger and I cherry-picked the changes that were approved, as the refactoring needs more effort to finish. We create a new PR for it: https://github.com/opf/openproject/pull/20371

myabc commented:
superseded by #20371. Will open separate PRs for eslint fixes and refactoring.
